File: android/src/main/java/app/capgo/plugin/health/background/BackgroundHealthApiClient.kt
package app.capgo.plugin.health.background

import android.util.Log
import app.capgo.plugin.health.HealthDataType
import com.getcapacitor.JSArray
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import org.json.JSONObject

class BackgroundHealthApiClient {
    fun fetchLastSyncMap(config: BackgroundSyncApiRequestConfig, subjectId: String): Map<HealthDataType, String> {
        val url = urlWithSubjectPath(config.url, subjectId)
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = CONNECT_TIMEOUT_MS
            readTimeout = READ_TIMEOUT_MS
            instanceFollowRedirects = true
        }
        config.headers.forEach { (key, value) ->
            connection.setRequestProperty(key, value)
        }
        return connection.useJsonConnection { response ->
            val json = JSONObject(response)
            val data = json.optJSONObject("data")
                ?: throw IllegalArgumentException("Missing data object in last sync response.")
            val items = data.optJSONArray("items")
                ?: throw IllegalArgumentException("Missing data.items in last sync response.")
            buildMap {
                for (i in 0 until items.length()) {
                    val row = items.optJSONObject(i) ?: continue
                    val key = row.optString("dataType", "")
                    if (key.isBlank()) {
                        continue
                    }
                    val dataType = HealthDataType.from(key)
                    if (dataType == null) {
                        Log.w(TAG, "Skipping unknown last-sync key (not in HealthDataType): $key")
                        continue
                    }
                    val timestamp = row.optString("lastSyncAt", "")
                    if (timestamp.isBlank()) {
                        throw IllegalArgumentException("Missing lastSyncAt for health data type: $key")
                    }
                    put(dataType, timestamp)
                }
            }
        }
    }

    fun uploadSamples(config: BackgroundSyncApiRequestConfig, subjectId: String, samples: JSArray) {
        val body = JSONObject().apply {
            put(
                "data",
                JSONObject().apply {
                    put("healthSubjectId", subjectId)
                    put("sourcePlatform", "HEALTH_CONNECT")
                    put("deliveredViaBackgroundSync", true)
                    put("samples", samples)
                }
            )
        }
        val connection = openConnection(config, "POST")
        connection.doOutput = true
        connection.setRequestProperty("Content-Type", "application/json")
        BufferedWriter(OutputStreamWriter(connection.outputStream)).use { writer ->
            writer.write(body.toString())
        }
        connection.useJsonConnection { }
    }

    /** MindMend zone-health: `{base}/{subjectId}` (path subject; not `?subjectId=`). */
    private fun urlWithSubjectPath(baseUrl: String, subjectId: String): String {
        val encoded = URLEncoder.encode(subjectId, StandardCharsets.UTF_8)
        return "${baseUrl.trimEnd('/')}/$encoded"
    }

    private fun openConnection(config: BackgroundSyncApiRequestConfig, method: String): HttpURLConnection {
        val connection = (URL(config.url).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = CONNECT_TIMEOUT_MS
            readTimeout = READ_TIMEOUT_MS
            instanceFollowRedirects = true
        }
        config.headers.forEach { (key, value) ->
            connection.setRequestProperty(key, value)
        }
        return connection
    }

    private inline fun <T> HttpURLConnection.useJsonConnection(block: (String) -> T): T {
        return try {
            val statusCode = responseCode
            val stream = if (statusCode in 200..299) inputStream else errorStream
            val responseBody = stream?.let { input ->
                BufferedReader(InputStreamReader(input)).use { reader ->
                    reader.readText()
                }
            }.orEmpty()
            if (statusCode !in 200..299) {
                throw IllegalStateException(
                    "Background sync API request failed with status $statusCode." +
                        if (responseBody.isNotBlank()) " Response: $responseBody" else ""
                )
            }
            block(responseBody)
        } finally {
            disconnect()
        }
    }

    companion object {
        private const val TAG = "BackgroundHealthApi"
        private const val CONNECT_TIMEOUT_MS = 30_000
        private const val READ_TIMEOUT_MS = 30_000
    }
}

File: ios/Sources/HealthPlugin/background/BackgroundHealthApiClient.swift
import Foundation

final class BackgroundHealthApiClient {
    private let session: URLSession
    private let connectTimeout: TimeInterval = 30
    private let readTimeout: TimeInterval = 30

    init(session: URLSession = .shared) {
        self.session = session
    }

    func fetchLastSyncMap(
        config: BackgroundSyncApiRequestConfig,
        subjectId: String,
        completion: @escaping (Result<[String: String], Error>) -> Void
    ) {
        let urlString = urlWithSubjectPath(baseUrl: config.url, subjectId: subjectId)
        guard let url = URL(string: urlString) else {
            completion(.failure(BackgroundHealthApiError.invalidUrl))
            return
        }

        var request = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: connectTimeout)
        request.httpMethod = "GET"
        applyHeaders(config.headers, to: &request)

        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            guard let http = response as? HTTPURLResponse else {
                completion(.failure(BackgroundHealthApiError.invalidResponse))
                return
            }
            guard (200 ... 299).contains(http.statusCode) else {
                let body = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
                completion(.failure(BackgroundHealthApiError.httpStatus(http.statusCode, body)))
                return
            }
            guard let data = data else {
                completion(.failure(BackgroundHealthApiError.emptyBody))
                return
            }
            do {
                completion(.success(try self.parseLastSyncMap(from: data)))
            } catch {
                completion(.failure(error))
            }
        }
        task.resume()
    }

    func uploadSamples(
        config: BackgroundSyncApiRequestConfig,
        subjectId: String,
        samples: [[String: Any]],
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        guard let url = URL(string: config.url) else {
            completion(.failure(BackgroundHealthApiError.invalidUrl))
            return
        }

        let body: [String: Any] = [
            "data": [
                "healthSubjectId": subjectId,
                "sourcePlatform": "APPLE_HEALTH",
                "deliveredViaBackgroundSync": true,
                "samples": samples
            ] as [String: Any]
        ]

        guard let httpBody = try? JSONSerialization.data(withJSONObject: body) else {
            completion(.failure(BackgroundHealthApiError.encodingFailed))
            return
        }

        var request = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: readTimeout)
        request.httpMethod = "POST"
        request.httpBody = httpBody
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        applyHeaders(config.headers, to: &request)

        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            guard let http = response as? HTTPURLResponse else {
                completion(.failure(BackgroundHealthApiError.invalidResponse))
                return
            }
            guard (200 ... 299).contains(http.statusCode) else {
                let body = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
                completion(.failure(BackgroundHealthApiError.httpStatus(http.statusCode, body)))
                return
            }
            completion(.success(()))
        }
        task.resume()
    }

    private func parseLastSyncMap(from data: Data) throws -> [String: String] {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObject = json["data"] as? [String: Any],
              let items = dataObject["items"] as? [[String: Any]] else {
            throw BackgroundHealthApiError.invalidPayload("Missing data.items in last sync response.")
        }

        var map: [String: String] = [:]
        for row in items {
            guard let dataType = row["dataType"] as? String, !dataType.isEmpty else { continue }
            guard HealthDataType(rawValue: dataType) != nil else {
                NSLog("[BackgroundHealthApi] Skipping unknown last-sync key: \(dataType)")
                continue
            }
            guard let lastSyncAt = row["lastSyncAt"] as? String, !lastSyncAt.isEmpty else {
                throw BackgroundHealthApiError.invalidPayload("Missing lastSyncAt for health data type: \(dataType)")
            }
            map[dataType] = lastSyncAt
        }
        return map
    }

    private func urlWithSubjectPath(baseUrl: String, subjectId: String) -> String {
        let trimmed = baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let encoded = subjectId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? subjectId
        return "\(trimmed)/\(encoded)"
    }

    private func applyHeaders(_ headers: [String: String], to request: inout URLRequest) {
        for (key, value) in headers {
            request.setValue(value, forHTTPHeaderField: key)
        }
    }
}

enum BackgroundHealthApiError: LocalizedError {
    case invalidUrl
    case invalidResponse
    case emptyBody
    case encodingFailed
    case httpStatus(Int, String)
    case invalidPayload(String)

    var errorDescription: String? {
        switch self {
        case .invalidUrl:
            return "Background sync API URL is invalid."
        case .invalidResponse:
            return "Background sync API response was invalid."
        case .emptyBody:
            return "Background sync API response body was empty."
        case .encodingFailed:
            return "Failed to encode background sync upload body."
        case let .httpStatus(code, body):
            if body.isEmpty {
                return "Background sync API request failed with status \(code)."
            }
            return "Background sync API request failed with status \(code). Response: \(body)"
        case let .invalidPayload(message):
            return message
        }
    }
}
